### DISCUSSION TITLE
Input Validation run separately for each dependent (depends not cached when input validation fails resulting in duplicated errors)

### DISCUSSION THREAD

**Discussion by ar45** [fastapi/fastapi#9005] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python

import logging
from pydantic import BaseModel
from fastapi import FastAPI, Depends, Request


logger = logging.Logger(__name__)


class UserModel(BaseModel):
    username: str
    name: str = None


app = FastAPI()


@Depends
def get_user_model_data(model: UserModel):
    print("Running function get_user_model_data")
    return model.dict(exclude_unset=True)


@Depends
def log_new_users(request: Request, data: dict = get_user_model_data):
    logger.info(f'Got request from {request.client.host} with username {model.username}')


@app.post('/users', dependencies=[log_new_users])
def create_account(data: dict = get_user_model_data):
    return {
        "status": "OK",
        "data": data,
    }

```

### Description

Above is a simple application with a /users endpoint using a pydantic model which requires a username in the post body.
when the model validation passes all is fine. But when model validation fails, the model is validated for each depends separately resulting in duplicated errors.

I would expect that it should be cached even when there are errors so it is validated only once.

* Run the app
* execute `curl -X POST "http://127.0.0.1:8000/api/users" -H  "Content-Type: application/json" -d '{"username":null, "name":"string"}'`
  it will result in duplicate error as follows
```
{
  "detail": [
    {
      "loc": [
        "body",
        "username"
      ],
      "msg": "none is not an allowed value",
      "type": "type_error.none.not_allowed"
    },
    {
      "loc": [
        "body",
        "username"
      ],
      "msg": "none is not an allowed value",
      "type": "type_error.none.not_allowed"
    }
  ]
}
```

### Environment

* OS: macOS:
* FastAPI Version 0.63.0:

* Python version: 3.9.1

**Comment by falkben** [comment]
Maybe you tried this before, but if you chain the `Depends` calls, you only get one error message. I undesrtand what you've described is undesired behavior, but it's not entirely unexpected. You have defined that dependency `get_user_model` twice, so if it errors, when it solves those it will show both errors.

If you want something with only one error, why not chain the dependencies, like so?

```python
import logging
from typing import Optional

from fastapi import Depends, FastAPI, Request
from pydantic import BaseModel

logger = logging.Logger(__name__)


class UserModel(BaseModel):
    username: str
    name: Optional[str] = None


app = FastAPI()


def get_user_model_data(model: UserModel):
    print("Running function get_user_model_data")
    return model.dict(exclude_unset=True)


def log_new_users(request: Request, data: dict = Depends(get_user_model_data)):
    logger.info(
        f"Got request from {request.client.host} with username {data.get('username')}"
    )
    return data


@app.post("/users")
def create_account(data: dict = Depends(log_new_users)):
    return {
        "status": "OK",
        "data": data,
    }
```

With that, when I hit the app with: `curl -X POST "http://127.0.0.1:8000/users" -H "Content-Type: application/json" -d '{"username":null, "name":"string"}'`

I get only one validation error: 

```json
{
    "detail": [
        {
            "loc": [
                "body",
                "username"
            ],
            "msg": "none is not an allowed value",
            "type": "type_error.none.not_allowed"
        }
    ]
}
```

**Comment by ar45** [comment]
@falkben yes, of course. 

In my case, one depends does not know about the other as these functions are defined in different scopes.


I submitted a pull request to cache the results
#2779
